### PR TITLE
ENH: nan_to_num keyword addition (was #9355)

### PR DESCRIPTION
**Pull request by kikocorreoso**
[This was #9355 ]

As I removed my local repo since #9355 was sent I send a new PR.

Sorry if this is not the canonical way. I tried to push to the old PR but I haven't found the way :disappointed: 

I hope this PR makes more sense than #9355 .

I've added tests to try to check if all is fine.

Please, let me know it all is fine or if it needs more work.

### CODE REVIEW

**Comment by ahaldane** [comment]
LGTM, seems like a good and pretty safe change.

It needs a release note. In `doc/release/1.17.0-notes.rst`, under Improvements, please add a section describing how `nan_to_num` now has new keyword args.

Are others (eg, @madphysicist , @eric-wieser who commented before, plus anyone else) ok with the parameter name change to `nan`, `posinf`, `neginf`? Kwd parameter names can be contentious

**Comment by ahaldane** [comment]
Some more minor notes:

We dropped the idea of a `dtype` kwd parameter brought up in #9355 . That's fine with me - we can always add it later in a future PR. We also dropped the idea of more careful treatment of complex types. That can also be left to other PRs.

I had a slight hesitation that the dispatcher should also return the nan, posinf and neginf parameters if not none. That could account for cases where the user supplied a scalar of a funny ducktype. But since those parameters are documented here to be scalars and not arrays, it seems better not to handle them in the dispatcher. (In my opinion, there are still unanswered questions about how ducktype'd scalars work for `__array_function__`)

**Comment by mattip** [comment]
Did this ever hit the mailing list?

**Comment by kikocorreoso** [comment]
Hi @ahaldane , 

Thanks for reviewing this. I added the release notes information. I hope all is fine.

Also thanks to @eric-wieser @madphysicist @mattip !!!

**Comment by ahaldane** [comment]
@kikocorreoso, looks good.

Following @mattip, I think it's worth posting a message to the mailing list about this PR, to make sure everyone is OK with the new functionality and keyword names. 

(Once this PR is merged, it's very hard for us to change the keywords later because of back-compatibility, so we want to get it right at the start!)

Just a short message will do, with a link to this PR and a brief description of the change and of the keyword names. The list is numpy-discussion@python.org, you may need to subscribe to it to post.

File: doc/release/1.17.0-notes.rst [hunk 1]
```diff
@@ -197,6 +197,12 @@ The boolean and integer types are incapable of storing ``np.nan`` and ``np.inf``
 which allows us to provide specialized ufuncs that are up to 250x faster than the current
 approach.
 
+New keywords added to ``np.nan_to_num``
+---------------------------------------
+``np.nan_to_num`` now accepts keywords ``nan``, ``posinf`` and ``neginf`` allowing the
```
Comment (ahaldane) [inline doc/release/1.17.0-notes.rst]: Circle-CI is complaining about "WARNING: Inline literal start-string without end-string." on this line, but I don't see the problem myself.

Comment (mattip) [reply doc/release/1.17.0-notes.rst]: it is the `s` on line 203, shpinx cannot parse
```
``code``s
```

**Comment by kikocorreoso** [comment]
@ahaldane @mattip 
I sent the email.

Also, I corrected the sphinx issue.

File: numpy/lib/tests/test_type_check.py [hunk 2]
```diff
@@ -360,6 +360,15 @@ def test_generic(self):
...
         assert_all(vals[2] > 1e10) and assert_all(np.isfinite(vals[2]))
         assert_equal(type(vals), np.ndarray)
+        
+        # perform the same tests but with nan, posinf and neginf keywords
+        with np.errstate(divide='ignore', invalid='ignore'):
+            vals = nan_to_num(np.array((-1., 0, 1))/0., 
+                              nan=10, posinf=20, neginf=30)
+        assert_all(vals[0] == 30) and assert_all(np.isfinite(vals[0]))
+        assert_(vals[1] == 10)
+        assert_all(vals[2] == 20) and assert_all(np.isfinite(vals[2]))
```
Comment (ahaldane) [inline numpy/lib/tests/test_type_check.py]: Can these three lines be replaced by `assert_equal(vals, [30, 10, 20])` ?

Comment (ahaldane) [reply numpy/lib/tests/test_type_check.py]: Oh I see, you were just following the pattern above. Still, I think it's better to use the shorter version if there is no problem.

File: numpy/lib/type_check.py [hunk 3]
```diff
@@ -444,10 +465,17 @@ def nan_to_num(x, copy=True):
...
     for d in dest:
-        _nx.copyto(d, 0.0, where=isnan(d))
-        _nx.copyto(d, maxf, where=isposinf(d))
-        _nx.copyto(d, minf, where=isneginf(d))
+        idx_nan = isnan(d)
+        idx_posinf = isposinf(d)
+        idx_neginf = isneginf(d)
+        _nx.copyto(d, nan, where=idx_nan)
+        _nx.copyto(d, maxf, where=idx_posinf)
+        _nx.copyto(d, minf, where=idx_neginf)
```
Comment (ahaldane) [inline numpy/lib/type_check.py]: I don't see the point of splitting these 3 lines into 6?

File: numpy/lib/tests/test_type_check.py [hunk 4]
```diff
@@ -371,26 +380,49 @@ def test_generic(self):
...
+        
+        # perform the same test but in-place
+        with np.errstate(divide='ignore', invalid='ignore'):
+            vals = np.array((-1., 0, 1))/0.
+        result = nan_to_num(vals, copy=False, nan=10, posinf=20, neginf=30)
+
+        assert_(result is vals)
+        assert_all(vals[0] == 30) and assert_all(np.isfinite(vals[0]))
+        assert_(vals[1] == 10)
+        assert_all(vals[2] == 20) and assert_all(np.isfinite(vals[2]))
```
Comment (ahaldane) [inline numpy/lib/tests/test_type_check.py]: Same here.

File: numpy/lib/tests/test_type_check.py (hunk 2 above)
Comment (kikocorreoso) [reply numpy/lib/tests/test_type_check.py]: Yes, I was following the previous code.

No problem to update this.

File: numpy/lib/type_check.py (hunk 3 above)
Comment (kikocorreoso) [reply numpy/lib/type_check.py]: This is to maintain what the user has chosen. If there is only this:

        ```
        _nx.copyto(d, nan, where=idx_nan)
        _nx.copyto(d, maxf, where=idx_posinf)
        _nx.copyto(d, minf, where=idx_neginf)
```
If in the keywords one chooses `nan=np.inf` and `posinf=-999` the following array:
`a = [np.nan, 10, np.inf]`
would be converted to:
`a = [-999, 10, -999]`
So the `nan` keyword would be overwritten as it is the first being executed. I think this wouldn't be the expected behavior. I would expect:
`a = [np.inf, 10, -999]`

This way it is necessary to create temporary boolean arrays but I think it makes more sense.

Comment (kikocorreoso) [reply numpy/lib/type_check.py]: Also, this is why I've added these tests:

https://github.com/numpy/numpy/pull/13219/files/f486ca24f20ca5d4de43191ef2d647c87c37ac1a#diff-c04fcf6a62bc1bda31fd3a1486786567R450

Comment (ahaldane) [reply numpy/lib/type_check.py]: Oh right, that makes sense, thanks

**Comment by ahaldane** [comment]
All right, looks good to me. Thanks @kikocorreoso , I'm going to merge now.

**Comment by kikocorreoso** [comment]
Thanks @ahaldane @mattip @eric-wieser @madphysicist

**Pull request merged**